src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,271 @@
+        os.path.join(dir, _SCALAFMT_CONF_FILENAME) for dir in source_dirs_with_ancestors
+    ]
+    config_files_snapshot = await Get(Snapshot, PathGlobs(config_file_globs))
+    config_files_set = set(config_files_snapshot.files)
+
+    config_files_for_source_dir: dict[str, str] = {}
+    for source_dir in source_dirs:
+        config_file = find_nearest_ancestor_config_file(config_files_set, source_dir)
+        if not config_file:
+            raise ValueError(
+                f"No scalafmt config file (`{_SCALAFMT_CONF_FILENAME}`) found for "
+                f"source directory '{source_dir}'"
+            )
+        config_files_for_source_dir[source_dir] = config_file
+
+    return ScalafmtConfigFiles(config_files_snapshot, FrozenDict(config_files_for_source_dir))
+
+
+@logging
+@rule(level=LogLevel.DEBUG)
+async def setup_scalafmt(
+    setup_request: SetupRequest,
+    tool: ScalafmtSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> Setup:
+    source_files, tool_classpath = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+        ),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(tool.resolved_lockfile(),),
+            ),
+        ),
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    config_files = await Get(
+        ScalafmtConfigFiles, GatherScalafmtConfigFilesRequest(source_files_snapshot)
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            [
+                source_files_snapshot.digest,
+                config_files.snapshot.digest,
+                tool_classpath.digest,
+                jdk_setup.digest,
+            ]
+        ),
+    )
+
+    # Partition the work by which source files share the same config file (regardless of directory).
+    source_files_by_config_file: dict[str, set[str]] = defaultdict(set)
+    for source_dir, files_in_source_dir in group_by_dir(source_files_snapshot.files).items():
+        config_file_for_source_dir = config_files.config_files_for_source_dir[source_dir]
+        source_files_by_config_file[config_file_for_source_dir].union(files_in_source_dir)
+
+    partitions = []
+    for config_file, files in source_files_by_config_file.items():
+        args = [
+            *jdk_setup.args(bash, tool_classpath.classpath_entries()),
+            "org.scalafmt.cli.Cli",
+            f"--config={config_file}",
+            "--non-interactive",
+        ]
+        if setup_request.check_only:
+            args.append("--list")
+        args.extend(sorted(files))
+
+        process = Process(
+            argv=args,
+            input_digest=input_digest,
```

Comment:
I think we want this to be the subset of files relevant to that partition. In particular, that's important so that changing `dirA` does not invalidate `dirB`.

To do that, you'll probably need to use `DigestSubset`. Will want to use that via `MultiGet` above the for loop. Also if there is only one partition, can shortcircuit and avoid calling `DigestSubset` for less wasted work.

Reply:
The latest commit uses `DigestSubset` as suggested and as required by bugs. The output digests cannot be merged when the same file has different contents (as expected when each partitions has been run through `fmt`).

I also broke out the partition setup into its own rule so I can now use `MultiGet` for all of the partitions.